File: clients/ui/src/state/endpointContractFixture.test.ts
/// <reference types="node" />

import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { create } from '@bufbuild/protobuf'
import { describe, expect, it } from 'vitest'
import { EndpointDaemonIdentitySchema, LocalDiscoveryCandidateSchema } from '../generated/remoteauthpb/remote_auth_pb'
import { decodeEndpointRegistryProjection, EndpointProjectionError, endpointRegistryMaxBytes, type EndpointConnectMode, type EndpointRegistryProjection, type EndpointRouteKind, type EndpointRouteProjection, type EndpointSource } from './endpointContract'

interface Candidate {
  source: EndpointSource
  identity: { device_id: string; device_fingerprint: string }
  suggested_label: string
  connect_mode?: EndpointConnectMode
  apply_client_policy?: boolean
  routes: Array<Record<string, unknown> & { route_id: string; kind: string }>
}

interface ConfirmedIdentityBinding {
  endpoint_id: string
  identity: { device_id: string; device_fingerprint: string }
}

interface ContractFixture {
  schema_version: number
  empty_registry: unknown
  valid_registry: unknown
  defaulted_managed_registry: unknown
  unknown_field_registry: unknown
  missing_field_registry: unknown
  wrong_type_registry: unknown
  whitespace_key_registry: unknown
  invalid_registry_cases: Array<{ name: string; registry: unknown; expected_error: string }>
  assembler: {
    initial_registry: unknown
    confirmed_identity_bindings: ConfirmedIdentityBinding[]
    commutative_candidates: Candidate[]
    identity_conflict_candidate: Candidate
    expected_endpoint_id: string
    expected_new_endpoint_id: string
    expected_label: string
    expected_connect_mode: EndpointConnectMode
    expected_route_ids: string[]
    expected_route_priorities: Record<string, number>
    expected_conflict_error: string
  }
  local_discovery_candidate: {
    claimed_device_id: string
    claimed_device_fingerprint: string
    address: string
    port: number
    protocol_version: number
    ttl_millis: number
    signature_bytes: number
  }
  oversize_bytes: number
}

describe('shared endpoint contract fixture', () => {
  it('agrees on strict parse, round-trip, size, identity conflict and import commutativity', async () => {
    const fixture = loadFixture()
    expect(fixture.schema_version).toBe(1)
    const empty = decodeEndpointRegistryProjection(JSON.stringify(fixture.empty_registry))
    expect(empty).toEqual({ version: 2, default: '', endpoints: {} })
    const registry = decodeEndpointRegistryProjection(JSON.stringify(fixture.valid_registry))
    expect(decodeEndpointRegistryProjection(JSON.stringify(registry))).toEqual(registry)
    const defaultedManaged = decodeEndpointRegistryProjection(JSON.stringify(fixture.defaulted_managed_registry))
    expect(defaultedManaged.endpoints.cloud?.routes.cloud?.relay_mode).toBe('auto')
    expect(() => decodeEndpointRegistryProjection(JSON.stringify(fixture.unknown_field_registry))).toThrowError(EndpointProjectionError)
    expect(() => decodeEndpointRegistryProjection(JSON.stringify(fixture.missing_field_registry))).toThrowError(EndpointProjectionError)
    expect(() => decodeEndpointRegistryProjection(JSON.stringify(fixture.wrong_type_registry))).toThrowError(EndpointProjectionError)
    expect(() => decodeEndpointRegistryProjection(JSON.stringify(fixture.whitespace_key_registry))).toThrowError(EndpointProjectionError)
    for (const testCase of fixture.invalid_registry_cases) {
      expect(() => decodeEndpointRegistryProjection(JSON.stringify(testCase.registry)), testCase.name).toThrowError(
        expect.objectContaining({ code: testCase.expected_error }),
      )
    }
    expect(() => decodeEndpointRegistryProjection('x'.repeat(fixture.oversize_bytes))).toThrowError(
      expect.objectContaining({ code: 'size_limit' }),
    )
    expect(fixture.oversize_bytes).toBe(endpointRegistryMaxBytes + 1)

    const nearby = fixture.local_discovery_candidate
    const discovery = create(LocalDiscoveryCandidateSchema, {
      claimedIdentity: create(EndpointDaemonIdentitySchema, {
        deviceId: nearby.claimed_device_id,
        deviceFingerprint: nearby.claimed_device_fingerprint,
      }),
      address: nearby.address,
      port: nearby.port,
      protocolVersion: nearby.protocol_version,
      announcementExpiresAtUnixNano: BigInt(Date.now() + nearby.ttl_millis) * 1_000_000n,
      announcementSignature: new Uint8Array(nearby.signature_bytes),
    })
    expect(discovery.announcementSignature).toHaveLength(64)
    expect(JSON.stringify(fixture.valid_registry)).not.toContain(discovery.address)

    const initial = decodeEndpointRegistryProjection(JSON.stringify(fixture.assembler.initial_registry))
    const forward = await assembleSequence(initial, fixture.assembler.commutative_candidates, fixture.assembler.confirmed_identity_bindings)
    const reverse = await assembleSequence(initial, [...fixture.assembler.commutative_candidates].reverse(), fixture.assembler.confirmed_identity_bindings)
    expect(reverse).toEqual(forward)
    expect(Object.keys(forward.endpoints)).toHaveLength(1)
    expect(forward.default).toBe(fixture.assembler.expected_endpoint_id)
    const fromEmpty = await assembleSequence(empty, fixture.assembler.commutative_candidates)
    expect(Object.keys(fromEmpty.endpoints)).toHaveLength(1)
    expect(fromEmpty.default).toBe(fixture.assembler.expected_new_endpoint_id)
    const endpoint = forward.endpoints[fixture.assembler.expected_endpoint_id]
    expect(endpoint?.label).toBe(fixture.assembler.expected_label)
    expect(endpoint?.connect_mode).toBe(fixture.assembler.expected_connect_mode)
    expect(Object.keys(endpoint?.routes ?? {}).sort()).toEqual(fixture.assembler.expected_route_ids)
    expect(Object.fromEntries(Object.entries(endpoint?.routes ?? {}).map(([routeId, route]) => [routeId, route.priority]))).toEqual(
      fixture.assembler.expected_route_priorities,
    )
    await expect(assembleSequence(forward, [fixture.assembler.identity_conflict_candidate])).rejects.toMatchObject({
      code: fixture.assembler.expected_conflict_error,
    })
  })
})

async function assembleSequence(
  initial: EndpointRegistryProjection,
  candidates: Candidate[],
  bindings: ConfirmedIdentityBinding[] = [],
): Promise<EndpointRegistryProjection> {
  let registry = structuredClone(initial)
  for (const [index, candidate] of candidates.entries()) {
    if (index === 0) registry = applyConfirmedIdentityBindings(registry, bindings, [candidate])
    registry = await assembleOne(registry, candidate)
  }
  return registry
}

async function assembleOne(registry: EndpointRegistryProjection, candidate: Candidate): Promise<EndpointRegistryProjection> {
  const next = structuredClone(registry)
  let matchId = ''
  for (const [endpointId, endpoint] of Object.entries(next.endpoints)) {
    if (!endpoint.device_id || !endpoint.device_fingerprint) continue
    if (endpoint.device_fingerprint === candidate.identity.device_fingerprint) {
      if (endpoint.device_id !== candidate.identity.device_id) throw { code: 'identity_conflict' }
      matchId = endpointId
    }
    if (endpoint.device_id === candidate.identity.device_id && endpoint.device_fingerprint !== candidate.identity.device_fingerprint) {
      throw { code: 'identity_conflict' }
    }
  }
  const endpointId = matchId || await deriveEndpointId(candidate.identity.device_fingerprint, next)
  const current = next.endpoints[endpointId]
  const incomingRank = sourceRank(candidate.source)
  const labelSource = current?.label_source ?? candidate.source
  const applyPolicy = candidate.source === 'local' || candidate.source === 'manual' || candidate.source === 'user'
    || candidate.source === 'share' && candidate.apply_client_policy === true
  const replaceLabel = current !== undefined && applyPolicy && labelSource !== 'user'
    && incomingRank >= sourceRank(labelSource) && candidate.suggested_label.trim() !== ''
  const label = current === undefined
    ? candidate.suggested_label.trim() || candidate.identity.device_id
    : replaceLabel ? candidate.suggested_label.trim() : current.label
  const routes = { ...(current?.routes ?? {}) }
  for (const candidateRoute of candidate.routes) {
    const existing = routes[candidateRoute.route_id]
    if (existing && existing.kind !== candidateRoute.kind) throw { code: 'route_conflict' }
    const { route_id: _routeId, ...portableRoute } = candidateRoute
    const incoming = {
      ...(portableRoute as unknown as NonNullable<typeof existing>),
      kind: candidateRoute.kind as EndpointRouteKind,
      enabled: candidateRoute.enabled !== false,
      source: candidate.source,
      policy_source: candidate.source,
    } satisfies EndpointRouteProjection
    let merged = existing
    if (!existing || incomingRank >= sourceRank(existing.source ?? 'manual')) {
      merged = existing ? withRoutePolicy(incoming, existing) : incoming
    }
    const forcePolicy = candidate.source === 'share' && candidate.apply_client_policy === true
    if (applyPolicy && merged && (forcePolicy || sourceRank(incoming.policy_source ?? candidate.source) >= sourceRank(existing?.policy_source ?? existing?.source ?? candidate.source))) {
      merged = withRoutePolicy(merged, incoming)
    }
    if (!existing && !applyPolicy) {
      merged = {
        ...(merged ?? incoming),
        manual_only: Object.values(routes).some((route) => route.enabled && !route.manual_only && route.priority !== undefined),
        policy_source: 'local',
      }
    }
    routes[candidateRoute.route_id] = merged ?? incoming
  }
  next.endpoints[endpointId] = {
    label,
    label_source: current === undefined || replaceLabel ? candidate.source : labelSource,
    device_id: candidate.identity.device_id,
    device_fingerprint: candidate.identity.device_fingerprint,
    enabled: current?.enabled ?? true,
    connect_mode: applyPolicy && candidate.connect_mode ? candidate.connect_mode : current?.connect_mode ?? 'on_demand',
    routes,
  }
  if (next.default === '') next.default = endpointId
  return decodeEndpointRegistryProjection(JSON.stringify(next))
}

function applyConfirmedIdentityBindings(
  registry: EndpointRegistryProjection,
  bindings: ConfirmedIdentityBinding[],
  candidates: Candidate[],
): EndpointRegistryProjection {
  if (bindings.length === 0) return registry
  const next = structuredClone(registry)
  const candidateIdentities = new Set(candidates.map((candidate) => identityKey(candidate.identity)))
  const seenEndpoints = new Set<string>()
  const seenIdentities = new Set<string>()
  for (const binding of [...bindings].sort((left, right) => left.endpoint_id.localeCompare(right.endpoint_id) || identityKey(left.identity).localeCompare(identityKey(right.identity)))) {
    const key = identityKey(binding.identity)
    if (!candidateIdentities.has(key)) throw { code: 'config_invalid' }
    if (seenEndpoints.has(binding.endpoint_id) || seenIdentities.has(key)) throw { code: 'identity_conflict' }
    seenEndpoints.add(binding.endpoint_id)
    seenIdentities.add(key)
    const endpoint = next.endpoints[binding.endpoint_id]
    if (!endpoint) throw { code: 'config_invalid' }
    for (const [endpointId, current] of Object.entries(next.endpoints)) {
      if (endpointId === binding.endpoint_id || !current.device_id || !current.device_fingerprint) continue
      if (current.device_id === binding.identity.device_id || current.device_fingerprint === binding.identity.device_fingerprint) {
        throw { code: 'identity_conflict' }
      }
    }
    if (endpoint.device_id || endpoint.device_fingerprint) {
      if (endpoint.device_id !== binding.identity.device_id || endpoint.device_fingerprint !== binding.identity.device_fingerprint) {
        throw { code: 'identity_conflict' }
      }
      continue
    }
    endpoint.device_id = binding.identity.device_id
    endpoint.device_fingerprint = binding.identity.device_fingerprint
  }
  return decodeEndpointRegistryProjection(JSON.stringify(next))
}

function identityKey(identity: { device_id: string; device_fingerprint: string }): string {
  return `${identity.device_id}\u0000${identity.device_fingerprint}`
}

function withRoutePolicy(base: EndpointRouteProjection, policy: EndpointRouteProjection): EndpointRouteProjection {
  const result = { ...base, enabled: policy.enabled }
  for (const key of ['manual_only', 'priority', 'policy_source'] as const) {
    if (policy[key] === undefined) delete result[key]
    else Object.assign(result, { [key]: policy[key] })
  }
  return result
}

async function deriveEndpointId(fingerprint: string, registry: EndpointRegistryProjection): Promise<string> {
  const digest = new Uint8Array(await crypto.subtle.digest('SHA-256', new TextEncoder().encode(fingerprint.trim())))
  const encoded = [...digest].map((value) => value.toString(16).padStart(2, '0')).join('')
  for (let length = 12; length <= encoded.length; length += 4) {
    const candidate = `daemon-${encoded.slice(0, length)}`
    if (!registry.endpoints[candidate]) return candidate
  }
  return `daemon-${encoded}`
}

function sourceRank(source: EndpointSource): number {
  return { lan: 0, cloud: 10, bootstrap: 20, local: 25, manual: 30, share: 40, user: 50 }[source]
}

function loadFixture(): ContractFixture {
  const path = resolve(process.cwd(), '../../shared/connection/testdata/endpoint-contract-v1.json')
  return JSON.parse(readFileSync(path, 'utf8')) as ContractFixture
}

File: clients/mobile/android/app/src/test/java/com/termx/app/endpoint/EndpointContractFixtureTest.kt
package com.termx.app.endpoint

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertThrows
import org.junit.Test
import com.google.protobuf.ByteString
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import termx.remote.auth.v1.RemoteAuth

/** EndpointContractFixtureTest 让 Android native 与 Go/TypeScript 读取同一份 registry/assembler testdata。 */
class EndpointContractFixtureTest {
    @Test
    fun sharedFixtureCoversStrictRoundTripStoreCommutativityAndConflict() {
        val fixture = loadFixture()
        assertEquals(1, fixture.getInt("schema_version"))

        val emptyRegistry = EndpointRegistryCodec.parseRegistry(fixture.getJSONObject("empty_registry"))
        assertEquals(SavedEndpointRegistry(), emptyRegistry)
        val blank = assertThrows(EndpointContractException::class.java) { EndpointRegistryCodec.decode(ByteArray(0)) }
        assertEquals("config_invalid", blank.code)

        val validPayload = fixture.getJSONObject("valid_registry").toString().toByteArray(StandardCharsets.UTF_8)
        val registry = EndpointRegistryCodec.decode(validPayload)
        val roundTripped = EndpointRegistryCodec.decode(EndpointRegistryCodec.encode(registry))
        assertEquals(registry, roundTripped)
        val defaultedManaged = EndpointRegistryCodec.parseRegistry(fixture.getJSONObject("defaulted_managed_registry"))
        assertEquals(EndpointRelayMode.AUTO, defaultedManaged.endpoints.getValue("cloud").routes.getValue("cloud").relayMode)

        val store = AndroidSavedEndpointRegistryStore(Files.createTempDirectory("termx-endpoints").resolve("endpoints.json").toFile())
        store.save(registry)
        assertEquals(registry, store.load())
        val updatedEndpoint = registry.endpoints.getValue(registry.defaultEndpointId).copy(label = "Updated Studio")
        val updatedRegistry = registry.copy(endpoints = registry.endpoints + (registry.defaultEndpointId to updatedEndpoint))
        store.save(updatedRegistry)
        assertEquals(updatedRegistry, store.load())
        val invalidRegistry = updatedRegistry.copy(
            endpoints = updatedRegistry.endpoints + (updatedRegistry.defaultEndpointId to updatedEndpoint.copy(routes = emptyMap())),
        )
        assertThrows(EndpointContractException::class.java) { store.save(invalidRegistry) }
        assertEquals(updatedRegistry, store.load())

        val unknown = assertThrows(EndpointContractException::class.java) {
            EndpointRegistryCodec.decode(fixture.getJSONObject("unknown_field_registry").toString().toByteArray(StandardCharsets.UTF_8))
        }
        assertEquals("config_invalid", unknown.code)
        val missingField = assertThrows(EndpointContractException::class.java) {
            EndpointRegistryCodec.decode(fixture.getJSONObject("missing_field_registry").toString().toByteArray(StandardCharsets.UTF_8))
        }
        assertEquals("config_invalid", missingField.code)
        val wrongType = assertThrows(EndpointContractException::class.java) {
            EndpointRegistryCodec.decode(fixture.getJSONObject("wrong_type_registry").toString().toByteArray(StandardCharsets.UTF_8))
        }
        assertEquals("config_invalid", wrongType.code)
        val whitespaceKey = assertThrows(EndpointContractException::class.java) {
            EndpointRegistryCodec.decode(fixture.getJSONObject("whitespace_key_registry").toString().toByteArray(StandardCharsets.UTF_8))
        }
        assertEquals("config_invalid", whitespaceKey.code)
        val invalidRegistryCases = fixture.getJSONArray("invalid_registry_cases")
        for (index in 0 until invalidRegistryCases.length()) {
            val testCase = invalidRegistryCases.getJSONObject(index)
            val failure = assertThrows(EndpointContractException::class.java) {
                EndpointRegistryCodec.decode(testCase.getJSONObject("registry").toString().toByteArray(StandardCharsets.UTF_8))
            }
            assertEquals(testCase.getString("name"), testCase.getString("expected_error"), failure.code)
        }
        val oversize = assertThrows(EndpointContractException::class.java) {
            EndpointRegistryCodec.decode(ByteArray(fixture.getInt("oversize_bytes")) { 'x'.code.toByte() })
        }
        assertEquals("size_limit", oversize.code)

        val discoveryFixture = fixture.getJSONObject("local_discovery_candidate")
        val discovery = RemoteAuth.LocalDiscoveryCandidate.newBuilder()
            .setClaimedIdentity(
                RemoteAuth.EndpointDaemonIdentity.newBuilder()
                    .setDeviceId(discoveryFixture.getString("claimed_device_id"))
                    .setDeviceFingerprint(discoveryFixture.getString("claimed_device_fingerprint")),
            )
            .setAddress(discoveryFixture.getString("address"))
            .setPort(discoveryFixture.getInt("port"))
            .setProtocolVersion(discoveryFixture.getInt("protocol_version"))
            .setAnnouncementExpiresAtUnixNano(System.currentTimeMillis() * 1_000_000 + discoveryFixture.getLong("ttl_millis") * 1_000_000)
            .setAnnouncementSignature(ByteString.copyFrom(ByteArray(discoveryFixture.getInt("signature_bytes"))))
            .build()
        assertEquals(64, discovery.announcementSignature.size())
        assertFalse(fixture.getJSONObject("valid_registry").toString().contains(discovery.address))

        val assemblerFixture = fixture.getJSONObject("assembler")
        val initial = EndpointRegistryCodec.parseRegistry(assemblerFixture.getJSONObject("initial_registry"))
        val candidateArray = assemblerFixture.getJSONArray("commutative_candidates")
        val candidates = (0 until candidateArray.length()).map { EndpointRegistryCodec.parseCandidate(candidateArray.getJSONObject(it)) }
        val bindingArray = assemblerFixture.getJSONArray("confirmed_identity_bindings")
        val bindings = (0 until bindingArray.length()).map { index ->
            val binding = bindingArray.getJSONObject(index)
            val identity = binding.getJSONObject("identity")
            ConfirmedEndpointIdentityBinding(
                endpointId = binding.getString("endpoint_id"),
                identity = EndpointDaemonIdentity(identity.getString("device_id"), identity.getString("device_fingerprint")),
            )
        }
        val forward = assembleSequence(initial, candidates, bindings)
        val reverse = assembleSequence(initial, candidates.reversed(), bindings)
        assertEquals(forward, reverse)
        assertEquals(1, forward.endpoints.size)
        assertEquals(assemblerFixture.getString("expected_endpoint_id"), forward.defaultEndpointId)
        val fromEmpty = assembleSequence(emptyRegistry, candidates, emptyList())
        assertEquals(1, fromEmpty.endpoints.size)
        assertEquals(assemblerFixture.getString("expected_new_endpoint_id"), fromEmpty.defaultEndpointId)

        val endpoint = forward.endpoints.getValue(assemblerFixture.getString("expected_endpoint_id"))
        assertEquals(assemblerFixture.getString("expected_label"), endpoint.label)
        assertEquals(EndpointConnectMode.parse(assemblerFixture.getString("expected_connect_mode")), endpoint.connectMode)
        val expectedRouteIds = assemblerFixture.getJSONArray("expected_route_ids").let { values ->
            (0 until values.length()).map(values::getString)
        }
        assertEquals(expectedRouteIds, endpoint.routes.keys.sorted())
        val expectedPriorities = assemblerFixture.getJSONObject("expected_route_priorities")
        endpoint.routes.forEach { (routeId, route) -> assertEquals(expectedPriorities.getInt(routeId), route.priority) }

        val conflictCandidate = EndpointRegistryCodec.parseCandidate(assemblerFixture.getJSONObject("identity_conflict_candidate"))
        val conflict = assertThrows(EndpointContractException::class.java) {
            AndroidEndpointAssembler.assemble(forward, listOf(conflictCandidate))
        }
        assertEquals(assemblerFixture.getString("expected_conflict_error"), conflict.code)
    }

    @Test
    fun emptyRegistryPublishesFirstEndpointAndReturnsDeterministicCredentialDescriptors() {
        val identity = EndpointDaemonIdentity("device-studio", "SHA256:studio")
        val result = AndroidEndpointAssembler.assemble(
            SavedEndpointRegistry(),
            listOf(
                AndroidEndpointCandidate(
                    source = EndpointSource.SHARE,
                    identity = identity,
                    routes = listOf(
                        SavedAccessRoute(
                            "ssh", EndpointRouteKind.SSH_STDIO, enabled = true, source = EndpointSource.SHARE,
                            policySource = EndpointSource.SHARE, host = "studio",
                        ),
                    ),
                    credentialDescriptors = listOf(
                        AndroidCredentialDescriptor("ssh-password", EndpointCredentialKind.SSH_PASSWORD, exportable = true),
                        AndroidCredentialDescriptor("ssh-key", EndpointCredentialKind.SSH_PRIVATE_KEY),
                        AndroidCredentialDescriptor("ssh-key", EndpointCredentialKind.SSH_PRIVATE_KEY),
                    ),
                ),
            ),
        )
        assertEquals(result.resolvedEndpointIds.single(), result.registry.defaultEndpointId)
        assertEquals(listOf("ssh-key", "ssh-password"), result.credentialDescriptors.map(AndroidCredentialDescriptor::descriptorId))

        val conflict = assertThrows(EndpointContractException::class.java) {
            AndroidEndpointAssembler.assemble(
                SavedEndpointRegistry(),
                listOf(
                    AndroidEndpointCandidate(
                        source = EndpointSource.SHARE,
                        identity = identity,
                        routes = listOf(
                            SavedAccessRoute(
                                "ssh", EndpointRouteKind.SSH_STDIO, enabled = true, source = EndpointSource.SHARE,
                                policySource = EndpointSource.SHARE, host = "studio",
                            ),
                        ),
                        credentialDescriptors = listOf(
                            AndroidCredentialDescriptor("credential", EndpointCredentialKind.SSH_PRIVATE_KEY),
                            AndroidCredentialDescriptor("credential", EndpointCredentialKind.SSH_PASSWORD),
                        ),
                    ),
                ),
            )
        }
        assertEquals("config_invalid", conflict.code)
    }

    @Test
    fun confirmedIdentityBindingPreservesSshEndpointAndUnconfirmedSharePreservesLabel() {
        val identity = EndpointDaemonIdentity("device-studio", "SHA256:studio")
        val ssh = SavedEndpoint(
            endpointId = "studio",
            label = "Local Studio",
            daemonIdentity = EndpointDaemonIdentity(),
            routes = mapOf(
                "ssh" to SavedAccessRoute(
                    routeId = "ssh",
                    kind = EndpointRouteKind.SSH_STDIO,
                    host = "studio-host",
                    remoteSocket = "auto",
                    source = EndpointSource.MANUAL,
                    policySource = EndpointSource.MANUAL,
                ),
            ),
        )
        val initial = SavedEndpointRegistry(defaultEndpointId = "studio", endpoints = mapOf("studio" to ssh))
        val bootstrap = AndroidEndpointCandidate(
            source = EndpointSource.BOOTSTRAP,
            identity = identity,
            routes = listOf(
                SavedAccessRoute(
                    routeId = "lan",
                    kind = EndpointRouteKind.DIRECT_TLS,
                    addresses = listOf("studio.local:41120"),
                    source = EndpointSource.BOOTSTRAP,
                    policySource = EndpointSource.BOOTSTRAP,
                ),
            ),
        )
        val bound = AndroidEndpointAssembler.assemble(
            initial,
            listOf(bootstrap),
            listOf(ConfirmedEndpointIdentityBinding("studio", identity)),
        ).registry
        assertEquals(setOf("studio"), bound.endpoints.keys)
        assertEquals(identity, bound.endpoints.getValue("studio").daemonIdentity)

        val unconfirmedShare = AndroidEndpointCandidate(
            source = EndpointSource.SHARE,
            identity = identity,
            suggestedLabel = "Shared Label",
            routes = listOf(
                SavedAccessRoute(
                    routeId = "ssh",
                    kind = EndpointRouteKind.SSH_STDIO,
                    host = "shared-host",
                    remoteSocket = "auto",
                    source = EndpointSource.SHARE,
                    policySource = EndpointSource.SHARE,
                ),
            ),
        )
        val shared = AndroidEndpointAssembler.assemble(bound, listOf(unconfirmedShare)).registry.endpoints.getValue("studio")
        assertEquals("Local Studio", shared.label)
        assertEquals("shared-host", shared.routes.getValue("ssh").host)

        val missingCandidate = assertThrows(EndpointContractException::class.java) {
            AndroidEndpointAssembler.assemble(
                initial,
                emptyList(),
                listOf(ConfirmedEndpointIdentityBinding("studio", identity)),
            )
        }
        assertEquals("config_invalid", missingCandidate.code)

        val invalidPolicy = initial.copy(
            endpoints = mapOf("studio" to ssh.copy(selectionPolicy = EndpointSelectionPolicy(1, false))),
        )
        val policyFailure = assertThrows(EndpointContractException::class.java) { EndpointRegistryCodec.encode(invalidPolicy) }
        assertEquals("config_invalid", policyFailure.code)
    }

    private fun assembleSequence(
        initial: SavedEndpointRegistry,
        candidates: List<AndroidEndpointCandidate>,
        bindings: List<ConfirmedEndpointIdentityBinding>,
    ): SavedEndpointRegistry {
        var registry = initial
        candidates.forEachIndexed { index, candidate ->
            registry = AndroidEndpointAssembler.assemble(
                registry,
                listOf(candidate),
                if (index == 0) bindings else emptyList(),
            ).registry
        }
        return registry
    }

    private fun loadFixture(): JSONObject {
        val payload = checkNotNull(javaClass.classLoader?.getResourceAsStream("endpoint-contract-v1.json")) {
            "shared endpoint fixture is missing"
        }.use { it.readBytes() }
        return JSONObject(payload.toString(StandardCharsets.UTF_8))
    }
}
